File: nars_scala/src/main/scala/nars/io/InferenceRecorder.scala
package nars.io

import java.awt.FileDialog
import java.io._
import nars.gui.InferenceWindow
//remove if not needed
import scala.collection.JavaConversions._

/**
 * Inference log, which record input/output of each inference step
 * jmv: make it an interface, with 2 implementations: GUI or batch, or apply MVC design pattern : append() is the event
 */
class InferenceRecorder {

  /**
   the display window
   */
  private var window: InferenceWindow = new InferenceWindow(this)

  /**
   whether to display
   */
  private var isReporting: Boolean = false

  /**
   the log file
   */
  private var logFile: PrintWriter = null

  /**
   * Initialize the window and the file
   */
  def init() {
    window.clear()
  }

  /**
   * Show the window
   */
  def show() {
    window.setVisible(true)
  }

  /**
   * Begin the display
   */
  def play() {
    isReporting = true
  }

  /**
   * Stop the display
   */
  def stop() {
    isReporting = false
  }

  /**
   * Add new text to display
   * @param s The line to be displayed
   */
  def append(s: String) {
    if (isReporting) {
      window.append(s)
    }
    if (logFile != null) {
      logFile.println(s)
    }
  }

  /**
   * Open the log file
   */
  def openLogFile() {
    val dialog = new FileDialog(null.asInstanceOf[FileDialog], "Inference Log", FileDialog.SAVE)
    dialog.setVisible(true)
    val directoryName = dialog.getDirectory
    val fileName = dialog.getFile
    try {
      logFile = new PrintWriter(new FileWriter(directoryName + fileName))
    } catch {
      case ex: IOException => println("i/o error: " + ex.getMessage)
    }
    window.switchBackground()
    window.setVisible(true)
  }

  /**
   * Close the log file
   */
  def closeLogFile() {
    logFile.close()
    logFile = null
    window.resetBackground()
  }

  /**
   * Check file logging
   * @return If the file logging is going on
   */
  def isLogging(): Boolean = (logFile != null)
}

File: nars_gui/nars/gui/InferenceRecorder.java
/*
 * InferenceRecorder.java
 *
 * Copyright (C) 2008  Pei Wang
 *
 * This file is part of Open-NARS.
 *
 * Open-NARS is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * Open-NARS is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Open-NARS.  If not, see <http://www.gnu.org/licenses/>.
 */
package nars.gui;

import java.awt.FileDialog;
import java.io.*;

import nars.io.IInferenceRecorder;

/**
 * Inference log, which record input/output of each inference step
 * interface with 1 implementation: GUI ( batch not implemented )
 */
public class InferenceRecorder implements IInferenceRecorder {

    /** the display window */
    private InferenceWindow window = new InferenceWindow(this);
    /** whether to display */
    private boolean isReporting = false;
    /** the log file */
    private PrintWriter logFile = null;

    @Override
	public void init() {
        window.clear();
    }

    @Override
	public void show() {
        window.setVisible(true);
    }

    @Override
	public void play() {
        isReporting = true;
    }

    @Override
	public void stop() {
        isReporting = false;
    }

    @Override
	public void append(String s) {
        if (isReporting) {
            window.append(s);
        }
        if (logFile != null) {
            logFile.println(s);
        }
    }

    @Override
	public void openLogFile() {
        FileDialog dialog = new FileDialog((FileDialog) null, "Inference Log", FileDialog.SAVE);
        dialog.setVisible(true);
        String directoryName = dialog.getDirectory();
        String fileName = dialog.getFile();
        try {
            logFile = new PrintWriter(new FileWriter(directoryName + fileName));
        } catch (IOException ex) {
            System.out.println("i/o error: " + ex.getMessage());
        }
        window.switchBackground();
        window.setVisible(true);
    }

    @Override
	public void closeLogFile() {
        logFile.close();
        logFile = null;
        window.resetBackground();
    }

    @Override
	public boolean isLogging() {
        return (logFile != null);
    }
}

